From 5c9c16b3bfdf0a32a54994e7ba5c0ee044222a7a Mon Sep 17 00:00:00 2001
From: Vincent Tabard <v@skyrock.com>
Date: Thu, 17 Apr 2014 17:45:31 +0200
Subject: [PATCH 1/4] Add an audio-channels property to decklinksrc in order to
 be able to fetch more than 2 channels of audio from the SDI signal (up to
 16).

---
 sys/decklink/gstdecklink.cpp    | 40 ++++++++++++++++++++++++++++++++++++++
 sys/decklink/gstdecklink.h      |  2 ++
 sys/decklink/gstdecklinksrc.cpp | 43 +++++++++++++++++++++++------------------
 sys/decklink/gstdecklinksrc.h   |  1 +
 4 files changed, 67 insertions(+), 19 deletions(-)

diff --git a/sys/decklink/gstdecklink.cpp b/sys/decklink/gstdecklink.cpp
index 848ff28..e3f6390 100644
--- a/sys/decklink/gstdecklink.cpp
+++ b/sys/decklink/gstdecklink.cpp
@@ -190,6 +190,46 @@ gst_decklink_mode_get_template_caps (void)
   return caps;
 }
 
+static GstStructure *
+gst_decklink_audio_channels_get_structure (int n_ch)
+{
+  return gst_structure_new ("audio/x-raw",
+      "format", G_TYPE_STRING, "S16LE",
+      "layout", G_TYPE_STRING, "interleaved",
+      "channels", G_TYPE_INT, n_ch,
+      "rate", G_TYPE_INT, 48000,
+      NULL);
+}
+
+GstCaps *
+gst_decklink_audio_channels_get_caps (int n_ch)
+{
+  GstCaps *caps;
+
+  caps = gst_caps_new_empty ();
+  gst_caps_append_structure (caps, gst_decklink_audio_channels_get_structure (n_ch));
+
+  return caps;
+}
+
+GstCaps *
+gst_decklink_audio_channels_get_template_caps (void)
+{
+  GstCaps *caps;
+  GstStructure *s;
+
+  caps = gst_caps_new_empty ();
+  // 2, 8, 16: only possibilities according to SDK doc
+  s = gst_decklink_audio_channels_get_structure (2);
+  gst_caps_append_structure (caps, s);
+  s = gst_decklink_audio_channels_get_structure (8);
+  gst_caps_append_structure (caps, s);
+  s = gst_decklink_audio_channels_get_structure (16);
+  gst_caps_append_structure (caps, s);
+
+  return caps;
+}
+
 typedef struct _Device Device;
 struct _Device {
   IDeckLink *decklink;
diff --git a/sys/decklink/gstdecklink.h b/sys/decklink/gstdecklink.h
index a204c78..380adc5 100644
--- a/sys/decklink/gstdecklink.h
+++ b/sys/decklink/gstdecklink.h
@@ -106,6 +106,8 @@ struct _GstDecklinkMode {
 const GstDecklinkMode * gst_decklink_get_mode (GstDecklinkModeEnum e);
 GstCaps * gst_decklink_mode_get_caps (GstDecklinkModeEnum e);
 GstCaps * gst_decklink_mode_get_template_caps (void);
+GstCaps * gst_decklink_audio_channels_get_caps (int n_ch);
+GstCaps * gst_decklink_audio_channels_get_template_caps (void);
 
 IDeckLink * gst_decklink_get_nth_device (int n);
 IDeckLinkInput * gst_decklink_get_nth_input (int n);
diff --git a/sys/decklink/gstdecklinksrc.cpp b/sys/decklink/gstdecklinksrc.cpp
index 7c2683a..8e3f57f 100644
--- a/sys/decklink/gstdecklinksrc.cpp
+++ b/sys/decklink/gstdecklinksrc.cpp
@@ -81,18 +81,11 @@ enum
   PROP_MODE,
   PROP_CONNECTION,
   PROP_AUDIO_INPUT,
-  PROP_DEVICE_NUMBER
+  PROP_DEVICE_NUMBER,
+  PROP_AUDIO_CHANNELS
 };
 
-static GstStaticPadTemplate gst_decklink_src_audio_src_template =
-GST_STATIC_PAD_TEMPLATE ("audiosrc",
-    GST_PAD_SRC,
-    GST_PAD_ALWAYS,
-    GST_STATIC_CAPS ("audio/x-raw, format=S16LE, channels=2, rate=48000, "
-        "layout=interleaved")
-    );
-
-/* the video source pad template is created on the fly */
+/* the audio + video source pad template is created on the fly */
 
 #define parent_class gst_decklink_src_parent_class
 G_DEFINE_TYPE (GstDecklinkSrc, gst_decklink_src, GST_TYPE_ELEMENT);
@@ -138,8 +131,15 @@ gst_decklink_src_class_init (GstDecklinkSrcClass * klass)
           (GParamFlags) (G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
               G_PARAM_CONSTRUCT)));
 
+  g_object_class_install_property (gobject_class, PROP_AUDIO_CHANNELS,
+      g_param_spec_int ("audio-channels", "Audio Channels", "Audio Channels to get from SDI signal (2, 8, 16)",
+          2, 16, 2,
+          (GParamFlags) (G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
+              G_PARAM_CONSTRUCT)));
+
   gst_element_class_add_pad_template (element_class,
-      gst_static_pad_template_get (&gst_decklink_src_audio_src_template));
+      gst_pad_template_new ("audiosrc", GST_PAD_SRC, GST_PAD_ALWAYS,
+          gst_decklink_audio_channels_get_template_caps ()));
 
   gst_element_class_add_pad_template (element_class,
       gst_pad_template_new ("videosrc", GST_PAD_SRC, GST_PAD_ALWAYS,
@@ -161,14 +161,12 @@ gst_decklink_src_init (GstDecklinkSrc * decklinksrc)
   gst_task_set_lock (decklinksrc->task, &decklinksrc->task_mutex);
 
   decklinksrc->audiosrcpad =
-      gst_pad_new_from_static_template (&gst_decklink_src_audio_src_template,
-      "audiosrc");
+      gst_pad_new_from_template (gst_element_class_get_pad_template
+      (GST_ELEMENT_CLASS (decklinksrc_class), "audiosrc"), "audiosrc");
   gst_pad_set_query_function (decklinksrc->audiosrcpad,
       GST_DEBUG_FUNCPTR (gst_decklink_src_audio_src_query));
   gst_element_add_pad (GST_ELEMENT (decklinksrc), decklinksrc->audiosrcpad);
 
-
-
   decklinksrc->videosrcpad =
       gst_pad_new_from_template (gst_element_class_get_pad_template
       (GST_ELEMENT_CLASS (decklinksrc_class), "videosrc"), "videosrc");
@@ -186,6 +184,7 @@ gst_decklink_src_init (GstDecklinkSrc * decklinksrc)
   decklinksrc->mode = GST_DECKLINK_MODE_NTSC;
   decklinksrc->connection = GST_DECKLINK_CONNECTION_SDI;
   decklinksrc->audio_connection = GST_DECKLINK_AUDIO_CONNECTION_AUTO;
+  decklinksrc->audio_channels = 2;
   decklinksrc->device_number = 0;
 
   decklinksrc->stop = FALSE;
@@ -238,6 +237,9 @@ gst_decklink_src_set_property (GObject * object, guint property_id,
     case PROP_DEVICE_NUMBER:
       decklinksrc->device_number = g_value_get_int (value);
       break;
+    case PROP_AUDIO_CHANNELS:
+      decklinksrc->audio_channels = g_value_get_int (value);
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
       break;
@@ -265,6 +267,9 @@ gst_decklink_src_get_property (GObject * object, guint property_id,
     case PROP_DEVICE_NUMBER:
       g_value_set_int (value, decklinksrc->device_number);
       break;
+    case PROP_AUDIO_CHANNELS:
+      g_value_set_int (value, decklinksrc->audio_channels);
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
       break;
@@ -509,11 +514,11 @@ gst_decklink_src_start (GstElement * element)
   }
 
   sample_depth = bmdAudioSampleType16bitInteger;
-  channels = 2;
+  channels = decklinksrc->audio_channels;
   ret = decklinksrc->input->EnableAudioInput (bmdAudioSampleRate48kHz,
       sample_depth, channels);
   if (ret != S_OK) {
-    GST_ERROR ("enable video input failed");
+    GST_ERROR ("enable audio input failed");
     return FALSE;
   }
 
@@ -856,8 +861,8 @@ gst_decklink_src_task (void *priv)
   if (gst_pad_is_linked (decklinksrc->audiosrcpad)) {
     n_samples = audio_frame->GetSampleFrameCount ();
     audio_frame->GetBytes (&data);
-    audio_buffer = gst_buffer_new_and_alloc (n_samples * 2 * 2);
-    gst_buffer_fill (audio_buffer, 0, data, n_samples * 2 * 2);
+    audio_buffer = gst_buffer_new_and_alloc (n_samples * decklinksrc->audio_channels * 2);
+    gst_buffer_fill (audio_buffer, 0, data, n_samples * decklinksrc->audio_channels * 2);
 
     GST_BUFFER_TIMESTAMP (audio_buffer) =
         gst_util_uint64_scale_int (decklinksrc->num_audio_samples * GST_SECOND,
diff --git a/sys/decklink/gstdecklinksrc.h b/sys/decklink/gstdecklinksrc.h
index f071b66..5623d64 100644
--- a/sys/decklink/gstdecklinksrc.h
+++ b/sys/decklink/gstdecklinksrc.h
@@ -82,6 +82,7 @@ struct _GstDecklinkSrc
   GstDecklinkModeEnum mode;
   GstDecklinkConnectionEnum connection;
   GstDecklinkAudioConnectionEnum audio_connection;
+  int audio_channels;
   int device_number;
 
 #ifdef _MSC_VER
-- 
1.9.1


From d139c8db58e8ed261afbce6951e0c442f24969b4 Mon Sep 17 00:00:00 2001
From: Vincent Tabard <v@skyrock.com>
Date: Thu, 17 Apr 2014 18:56:12 +0200
Subject: [PATCH 2/4] Push new caps on audio-channels decision

---
 sys/decklink/gstdecklinksrc.cpp | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/sys/decklink/gstdecklinksrc.cpp b/sys/decklink/gstdecklinksrc.cpp
index 8e3f57f..789155a 100644
--- a/sys/decklink/gstdecklinksrc.cpp
+++ b/sys/decklink/gstdecklinksrc.cpp
@@ -718,10 +718,7 @@ gst_decklink_src_send_initial_events (GstDecklinkSrc * src)
 
   /* caps */
   gst_pad_push_event (src->audiosrcpad,
-      gst_event_new_caps (gst_caps_new_simple ("audio/x-raw",
-          "format", G_TYPE_STRING, "S16LE", "channels", G_TYPE_INT, 2,
-          "rate", G_TYPE_INT, 48000, "layout", G_TYPE_STRING, "interleaved",
-          NULL)));
+      gst_event_new_caps (gst_decklink_audio_channels_get_caps (src->audio_channels)));
 
   gst_pad_push_event (src->videosrcpad,
       gst_event_new_caps (gst_decklink_mode_get_caps (src->mode)));
-- 
1.9.1


From ab94f55cbcfcf014f4ec6accf4c7e30141b493b0 Mon Sep 17 00:00:00 2001
From: Vincent Tabard <v@skyrock.com>
Date: Thu, 17 Apr 2014 22:39:12 +0200
Subject: [PATCH 3/4] Add the channel-mask to the audio caps so that the
 downstream elements don't complain. Fixme: missing for 16-channel audio.

---
 sys/decklink/gstdecklink.cpp | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/sys/decklink/gstdecklink.cpp b/sys/decklink/gstdecklink.cpp
index e3f6390..82f756d 100644
--- a/sys/decklink/gstdecklink.cpp
+++ b/sys/decklink/gstdecklink.cpp
@@ -22,6 +22,7 @@
 #endif
 
 #include <gst/gst.h>
+#include <gst/audio/audio-channels.h>
 #include "gstdecklink.h"
 #include "gstdecklinksrc.h"
 #include "gstdecklinksink.h"
@@ -193,12 +194,39 @@ gst_decklink_mode_get_template_caps (void)
 static GstStructure *
 gst_decklink_audio_channels_get_structure (int n_ch)
 {
-  return gst_structure_new ("audio/x-raw",
+  GstStructure * s = gst_structure_new ("audio/x-raw",
       "format", G_TYPE_STRING, "S16LE",
       "layout", G_TYPE_STRING, "interleaved",
       "channels", G_TYPE_INT, n_ch,
       "rate", G_TYPE_INT, 48000,
       NULL);
+
+  guint64 channel_mask = 0;
+  switch (n_ch) {
+    /* FIXME: need 16 channels layout */
+    case 8:
+      channel_mask =
+        GST_AUDIO_CHANNEL_POSITION_MASK (FRONT_LEFT) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (FRONT_RIGHT) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (REAR_LEFT) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (REAR_RIGHT) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (FRONT_CENTER) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (LFE1) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (SIDE_LEFT) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (SIDE_RIGHT);
+      break;
+    case 2:
+      channel_mask =
+        GST_AUDIO_CHANNEL_POSITION_MASK (FRONT_LEFT) |
+        GST_AUDIO_CHANNEL_POSITION_MASK (FRONT_RIGHT);
+      break;
+    default:
+      channel_mask = 0;
+      break;
+  }
+  gst_structure_set (s, "channel-mask", GST_TYPE_BITMASK, channel_mask, NULL);
+  
+  return s;
 }
 
 GstCaps *
-- 
1.9.1


From 7510611fcb213de7c94d72f1f1eec2c00a036ec2 Mon Sep 17 00:00:00 2001
From: Vincent Tabard <v@skyrock.com>
Date: Fri, 18 Apr 2014 07:58:26 +0200
Subject: [PATCH 4/4] Make the int an enum, since we only have 3 allowed values

---
 sys/decklink/gstdecklink.cpp    | 30 +++++++++++++++++++++++++-----
 sys/decklink/gstdecklink.h      | 10 +++++++++-
 sys/decklink/gstdecklinksrc.cpp | 15 +++++++++------
 sys/decklink/gstdecklinksrc.h   |  2 +-
 4 files changed, 44 insertions(+), 13 deletions(-)

diff --git a/sys/decklink/gstdecklink.cpp b/sys/decklink/gstdecklink.cpp
index 82f756d..69e3591 100644
--- a/sys/decklink/gstdecklink.cpp
+++ b/sys/decklink/gstdecklink.cpp
@@ -111,6 +111,26 @@ gst_decklink_audio_connection_get_type (void)
   return (GType) id;
 }
 
+GType
+gst_decklink_audio_channels_get_type (void)
+{
+  static gsize id = 0;
+  static const GEnumValue channel_numbers[] = {
+    {GST_DECKLINK_AUDIO_CHANNELS_2, "2ch", "2 audio channels"},
+    {GST_DECKLINK_AUDIO_CHANNELS_8, "8ch", "8 audio channels"},
+    {GST_DECKLINK_AUDIO_CHANNELS_16, "16ch", "16 audio channels"},
+    {0, NULL, NULL}
+  };
+
+  if (g_once_init_enter (&id)) {
+    GType tmp =
+        g_enum_register_static ("GstDecklinkAudioChannels", channel_numbers);
+    g_once_init_leave (&id, tmp);
+  }
+
+  return (GType) id;
+}
+
 #define NTSC 10, 11, false, false
 #define PAL 12, 11, true, false
 #define HD 1, 1, false, true
@@ -192,7 +212,7 @@ gst_decklink_mode_get_template_caps (void)
 }
 
 static GstStructure *
-gst_decklink_audio_channels_get_structure (int n_ch)
+gst_decklink_audio_channels_get_structure (GstDecklinkAudioChannelsEnum n_ch)
 {
   GstStructure * s = gst_structure_new ("audio/x-raw",
       "format", G_TYPE_STRING, "S16LE",
@@ -230,7 +250,7 @@ gst_decklink_audio_channels_get_structure (int n_ch)
 }
 
 GstCaps *
-gst_decklink_audio_channels_get_caps (int n_ch)
+gst_decklink_audio_channels_get_caps (GstDecklinkAudioChannelsEnum n_ch)
 {
   GstCaps *caps;
 
@@ -248,11 +268,11 @@ gst_decklink_audio_channels_get_template_caps (void)
 
   caps = gst_caps_new_empty ();
   // 2, 8, 16: only possibilities according to SDK doc
-  s = gst_decklink_audio_channels_get_structure (2);
+  s = gst_decklink_audio_channels_get_structure (GST_DECKLINK_AUDIO_CHANNELS_2);
   gst_caps_append_structure (caps, s);
-  s = gst_decklink_audio_channels_get_structure (8);
+  s = gst_decklink_audio_channels_get_structure (GST_DECKLINK_AUDIO_CHANNELS_8);
   gst_caps_append_structure (caps, s);
-  s = gst_decklink_audio_channels_get_structure (16);
+  s = gst_decklink_audio_channels_get_structure (GST_DECKLINK_AUDIO_CHANNELS_16);
   gst_caps_append_structure (caps, s);
 
   return caps;
diff --git a/sys/decklink/gstdecklink.h b/sys/decklink/gstdecklink.h
index 380adc5..9f2bd3d 100644
--- a/sys/decklink/gstdecklink.h
+++ b/sys/decklink/gstdecklink.h
@@ -89,6 +89,14 @@ typedef enum {
 #define GST_TYPE_DECKLINK_AUDIO_CONNECTION (gst_decklink_audio_connection_get_type ())
 GType gst_decklink_audio_connection_get_type (void);
 
+typedef enum {
+  GST_DECKLINK_AUDIO_CHANNELS_2 = 2,
+  GST_DECKLINK_AUDIO_CHANNELS_8 = 8,
+  GST_DECKLINK_AUDIO_CHANNELS_16 = 16
+} GstDecklinkAudioChannelsEnum;
+#define GST_TYPE_DECKLINK_AUDIO_CHANNELS (gst_decklink_audio_channels_get_type ())
+GType gst_decklink_audio_channels_get_type (void);
+
 typedef struct _GstDecklinkMode GstDecklinkMode;
 struct _GstDecklinkMode {
   BMDDisplayMode mode;
@@ -106,7 +114,7 @@ struct _GstDecklinkMode {
 const GstDecklinkMode * gst_decklink_get_mode (GstDecklinkModeEnum e);
 GstCaps * gst_decklink_mode_get_caps (GstDecklinkModeEnum e);
 GstCaps * gst_decklink_mode_get_template_caps (void);
-GstCaps * gst_decklink_audio_channels_get_caps (int n_ch);
+GstCaps * gst_decklink_audio_channels_get_caps (GstDecklinkAudioChannelsEnum n_ch);
 GstCaps * gst_decklink_audio_channels_get_template_caps (void);
 
 IDeckLink * gst_decklink_get_nth_device (int n);
diff --git a/sys/decklink/gstdecklinksrc.cpp b/sys/decklink/gstdecklinksrc.cpp
index 789155a..cf71750 100644
--- a/sys/decklink/gstdecklinksrc.cpp
+++ b/sys/decklink/gstdecklinksrc.cpp
@@ -132,8 +132,10 @@ gst_decklink_src_class_init (GstDecklinkSrcClass * klass)
               G_PARAM_CONSTRUCT)));
 
   g_object_class_install_property (gobject_class, PROP_AUDIO_CHANNELS,
-      g_param_spec_int ("audio-channels", "Audio Channels", "Audio Channels to get from SDI signal (2, 8, 16)",
-          2, 16, 2,
+      g_param_spec_enum ("audio-channels", "Audio Channels",
+          "Audio Channels to get from SDI signal (2, 8, 16)",
+          GST_TYPE_DECKLINK_AUDIO_CHANNELS,
+          GST_DECKLINK_AUDIO_CHANNELS_2,
           (GParamFlags) (G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
               G_PARAM_CONSTRUCT)));
 
@@ -184,7 +186,7 @@ gst_decklink_src_init (GstDecklinkSrc * decklinksrc)
   decklinksrc->mode = GST_DECKLINK_MODE_NTSC;
   decklinksrc->connection = GST_DECKLINK_CONNECTION_SDI;
   decklinksrc->audio_connection = GST_DECKLINK_AUDIO_CONNECTION_AUTO;
-  decklinksrc->audio_channels = 2;
+  decklinksrc->audio_channels = GST_DECKLINK_AUDIO_CHANNELS_2;
   decklinksrc->device_number = 0;
 
   decklinksrc->stop = FALSE;
@@ -238,7 +240,8 @@ gst_decklink_src_set_property (GObject * object, guint property_id,
       decklinksrc->device_number = g_value_get_int (value);
       break;
     case PROP_AUDIO_CHANNELS:
-      decklinksrc->audio_channels = g_value_get_int (value);
+      decklinksrc->audio_channels =
+          (GstDecklinkAudioChannelsEnum) g_value_get_enum (value);
       break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
@@ -268,7 +271,7 @@ gst_decklink_src_get_property (GObject * object, guint property_id,
       g_value_set_int (value, decklinksrc->device_number);
       break;
     case PROP_AUDIO_CHANNELS:
-      g_value_set_int (value, decklinksrc->audio_channels);
+      g_value_set_enum (value, decklinksrc->audio_channels);
       break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
@@ -514,7 +517,7 @@ gst_decklink_src_start (GstElement * element)
   }
 
   sample_depth = bmdAudioSampleType16bitInteger;
-  channels = decklinksrc->audio_channels;
+  channels = (int) decklinksrc->audio_channels;
   ret = decklinksrc->input->EnableAudioInput (bmdAudioSampleRate48kHz,
       sample_depth, channels);
   if (ret != S_OK) {
diff --git a/sys/decklink/gstdecklinksrc.h b/sys/decklink/gstdecklinksrc.h
index 5623d64..9f890e3 100644
--- a/sys/decklink/gstdecklinksrc.h
+++ b/sys/decklink/gstdecklinksrc.h
@@ -82,7 +82,7 @@ struct _GstDecklinkSrc
   GstDecklinkModeEnum mode;
   GstDecklinkConnectionEnum connection;
   GstDecklinkAudioConnectionEnum audio_connection;
-  int audio_channels;
+  GstDecklinkAudioChannelsEnum audio_channels;
   int device_number;
 
 #ifdef _MSC_VER
-- 
1.9.1

